### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by Joepena**

### CODE REVIEW

File: actions/app.go [hunk 1]
```diff
@@ -8,8 +8,12 @@ import (
 	"github.com/unrolled/secure"
 
 	"github.com/gobuffalo/x/sessions"
+	"github.com/joepena/monsters/models"
```
Comment (villejacob) [inline actions/app.go]: Can we make this the relative path Monsters/models?

File: actions/auth.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	f := c.Request().Form
+
+	err := validateCreateUserReqForm(f)
+	if err != nil {
+		return err
+	}
+
+	token, err := getAuthToken(c)
+	if err != nil {
+		log.Error(err)
```
Comment (villejacob) [inline actions/auth.go]: Which errors do you want logged?

File: actions/auth.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	err = u.Create()
+
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	c.Render(201, render.JSON(map[string]string{"token": u.AuthToken, "email": u.Email, "userId": u.ID}))
+
+	return nil
```
Comment (villejacob) [inline actions/auth.go]: Return the c.Render(...) instead of separate nil

File: actions/auth.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+func validateCreateUserReqForm(values url.Values) error {
+	email := values.Get("email")
+	password := values.Get("password")
+
+	if email == "" {
+		return errors.New("Empty email was provided.")
```
Comment (villejacob) [inline actions/auth.go]: Errors should be lowercase with no terminating punctuation

File: actions/auth.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	validUser := u.Authenticate()
+	if !validUser {
+		err := errors.New("invalid credentials")
+		log.Error(err)
+		return err
+	}
+
+	c.Render(201, render.JSON(map[string]string{"token": u.AuthToken, "userId": u.ID}))
+
+	return nil
```
Comment (villejacob) [inline actions/auth.go]: See prev comment

File: models/models.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+		MaxBytes: capacity,
+	}
+	err := db.session.DB(dbName).C(collectionName).Create(collectionInfo)
+	if err != nil {
+		log.WithField("Function", "CreateCappedCollection").Error(err)
+	}
+}
+
+func (db *DB) GetUserById(id string) (User, error) {
+	collection := db.session.DB("auth").C("users")
```
Comment (villejacob) [inline models/models.go]: Is it more "go" to name this 'c' instead of 'collection'?

File: models/user.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+func (u *User) Create() error {
+	collection := GetDBInstance().session.DB("auth").C("users")
+	var testU User
+
+	//check if email is already in DB
+	u.Email = strings.ToLower(u.Email)
+	_ = collection.Find(bson.M{"email": strings.ToLower(u.Email),}).One(&testU)
+
+	if testU.Email == u.Email {
+		return errors.New("Email is already in use.")
```
Comment (villejacob) [inline models/user.go]: See prev comment

File: models/user.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+// Create wraps up the pattern of encrypting the password and
+// running validations.
+func (u *User) Create() error {
+	collection := GetDBInstance().session.DB("auth").C("users")
+	var testU User
+
+	//check if email is already in DB
+	u.Email = strings.ToLower(u.Email)
+	_ = collection.Find(bson.M{"email": strings.ToLower(u.Email),}).One(&testU)
```
Comment (villejacob) [inline models/user.go]: I think you can call this without assigning the _

File: actions/auth.go (hunk 2 above)
Comment (Joepena) [reply actions/auth.go]: These are general debug errors. I've been using the general rule of thumb of logging all errors so its easy to catch bugs. What do you think of that?

File: actions/auth.go (hunk 3 above)
Comment (Joepena) [reply actions/auth.go]: good idea!

File: models/models.go (hunk 6 above)
Comment (Joepena) [reply models/models.go]: yeah go takes its naming conventions from c so it makes sense to do that. Im just a java boy so i have a bad naming habit ill fix it. ;)

**Pull request merged**